README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ror

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<img src='man/figures/logo.png' align="right" height="139" />

<!-- badges: end -->

`ror4r` is an R wrapper for the [Research Organization Registry
(ROR)](#https://ror.org) API. ROR is a registry of more than 100,000
research organisations across the globe. It holds data on location,
aliases and type of organisation - plus much more.

Currently, the `ror4r` package lets you return a list of organisations
(`get_org_list()`) or details for a specific organisation if you know
its ID (`get_single_org()`) via the API. If you know IDs from other
registries (eg, GRID, UKPRN), you can use `get_id_map()` to lookup other
IDs for the organisation ROR has stored.

You may find the [API
documentation](https://ror.readme.io/docs/rest-api), or the [reference
site](http://shanj90.github.io/ror4r) helpful in using this wrapper. The
[ROR API GitHub repo](https://github.com/ror-community/ror-api) may also
be of use.

In addition to the API interactions, you can also use `read_data_dump()`
to extract a dataframe of all the data from the system.

## Installation

You can install the development version of ror like so:

``` r
remotes::install_github("shanej90/ror4r")
```

## Example

The below example shows you how to return details for a specific
organisation:

``` r
library(ror4r)

data <- get_single_org(id = "03yghzc09") #University of Exeter
#> No encoding supplied: defaulting to UTF-8.

data
#> $id
#> [1] "https://ror.org/03yghzc09"
#> 
#> $name
#> [1] "University of Exeter"
#> 
#> $email_address
#> NULL
#> 
#> $ip_addresses
#> list()
#> 
#> $established
#> [1] 1955
#> 
#> $types
#> [1] "Education"
#> 
#> $relationships
#>                                                       label    type
#> 1                                        Derriford Hospital Related
#> 2               Peninsula College of Medicine and Dentistry Related
#> 3                           Royal Devon and Exeter Hospital Related
#> 4    Wellcome Centre of Cultures and Environments of Health   Child
#> 5 GW4 Facility for High-Resolution Electron Cryo-Microscopy Related
#> 6                                                       GW4 Related
#> 7                           MRC Centre for Medical Mycology   Child
#>                          id
#> 1 https://ror.org/00v5h4y49
#> 2 https://ror.org/04dtfyh05
#> 3 https://ror.org/03jrh3t05
#> 4 https://ror.org/00rbqbc98
#> 5 https://ror.org/022p86748
#> 6 https://ror.org/000vekr11
#> 7 https://ror.org/00vbzva31
#> 
#> $addresses
#>       lat      lng state state_code   city geonames_city.id geonames_city.city
#> 1 50.7236 -3.52751    NA         NA Exeter          2649808             Exeter
#>   geonames_city.geonames_admin1.name geonames_city.geonames_admin1.id
#> 1                            England                          6269131
#>   geonames_city.geonames_admin1.ascii_name geonames_city.geonames_admin1.code
#> 1                                  England                             GB.ENG
#>   geonames_city.geonames_admin2.name geonames_city.geonames_admin2.id
#> 1                              Devon                          2651292
#>   geonames_city.geonames_admin2.ascii_name geonames_city.geonames_admin2.code
#> 1                                    Devon                          GB.ENG.D4
#>                  geonames_city.license.attribution
#> 1 Data from geonames.org under a CC-BY 3.0 license
#>                 geonames_city.license.license geonames_city.nuts_level1.name
#> 1 http://creativecommons.org/licenses/by/3.0/                             NA
#>   geonames_city.nuts_level1.code geonames_city.nuts_level2.name
#> 1                             NA                             NA
#>   geonames_city.nuts_level2.code geonames_city.nuts_level3.name
#> 1                             NA                             NA
#>   geonames_city.nuts_level3.code postcode primary line country_geonames_id
#> 1                             NA       NA   FALSE   NA             2635167
#> 
#> $links
#> [1] "http://www.exeter.ac.uk/"
#> 
#> $aliases
#> list()
#> 
#> $acronyms
#> list()
#> 
#> $status
#> [1] "active"
#> 
#> $wikipedia_url
#> [1] "http://en.wikipedia.org/wiki/University_of_Exeter"
#> 
#> $labels
#>                label iso639
#> 1 Prifysgol Caerwysg     cy
#> 
#> $country
#> $country$country_name
#> [1] "United Kingdom"
#> 
#> $country$country_code
#> [1] "GB"
#> 
#> 
#> $external_ids
#> $external_ids$ISNI
#> $external_ids$ISNI$preferred
#> NULL
#> 
#> $external_ids$ISNI$all
#> [1] "0000 0004 1936 8024"
#> 
#> 
#> $external_ids$FundRef
#> $external_ids$FundRef$preferred
#> [1] "501100000737"
#> 
#> $external_ids$FundRef$all
#> [1] "501100000737" "501100000604"
#> 
#> 
#> $external_ids$HESA
#> $external_ids$HESA$preferred
#> NULL
#> 
#> $external_ids$HESA$all
#> [1] "0119"
#> 
#> 
#> $external_ids$UCAS
#> $external_ids$UCAS$preferred
#> NULL
#> 
#> $external_ids$UCAS$all
#> [1] "E84"
#> 
#> 
#> $external_ids$UKPRN
#> $external_ids$UKPRN$preferred
#> NULL
#> 
#> $external_ids$UKPRN$all
#> [1] "10007792"
#> 
#> 
#> $external_ids$OrgRef
#> $external_ids$OrgRef$preferred
#> NULL
#> 
#> $external_ids$OrgRef$all
#> [1] "33719893"
#> 
#> 
#> $external_ids$Wikidata
#> $external_ids$Wikidata$preferred
#> NULL
#> 
#> $external_ids$Wikidata$all
#> [1] "Q1414861"
#> 
#> 
#> $external_ids$GRID
#> $external_ids$GRID$preferred
#> [1] "grid.8391.3"
#> 
#> $external_ids$GRID$all
#> [1] "grid.8391.3"
```

## Future developments

At the moment, the package purely makes calls to the API and returns the
data in an appropriate format. Future plans for the package will further
process the data to meet specific goals. At the moment you’ll need to do
this outside of the package using your own understanding of the data.

Feel free to fork if you’d like to have a go yourself.

## Logo attribution

Icon from [www.flaticon.com](https://www.flaticon.com).


NEWS.md
# ror4r 0.1.2

* Updated functions to work with [December 2022 API change](https://ror.readme.io/changelog/2022-12-01-organization-status-changes) to allow filtering by status.
* Added a `NEWS.md` file to track changes to the package.
* Updated package name to `ror4r` to avoid CRAN conflict.


tests/testthat.R
library(testthat)
library(ror4r)

test_check("ror4r")


tests/testthat/test-get_org_list.R
#s3 classes----------------------------------------------

#test that you get a dataframe, assuming no errors
testthat::expect_s3_class(
  get_org_list(.page_numbers = 1),
  "data.frame"
)

#error handling------------------------------------------------------

#test that incorrect org type gives an error
testthat::test_that(
  "wring org_type gives and error",
  expect_error(
    get_org_list(.org_type = "blah"),
    "`.org_type` must be (if used) one of: 'Education', 'Healthcare', 'Company', 'Archive', 'Nonprofit', 'Government', 'Facility', or 'Other'.",
    fixed = T
  )
)

#>1 value for a single parameter
testthat::test_that(
  "that you can only enter one country code",
  expect_error(
    get_org_list(.country_code = c("GB", "US")),
    "You can only enter a single filter value for each parameter.",
    fixed = T
  )
)

#you can only enter a country code or name, not both
testthat::test_that(
  "that you can only enter one country code",
  expect_error(
    get_org_list(.country_code = "FR", .country_name = "France"),
    "You can only filter one of `.country_code` or `.country_name`.",
    fixed = T
  )
)

#that if using affiliation, you can't enter other filters/terms, except page numbers
testthat::test_that(
  "that you can only enter one country code",
  expect_error(
    get_org_list(.affiliation = "University of Exeter", .org_type = "Education"),
    "If using `.affiliation` you can not use `.search_terms` or any filters.",
    fixed = T
  )
)


tests/testthat/test-get_single_org.R
#check that you get a list
testthat::expect_type(
  get_single_org(id = "03yghzc09"),
  "list"
)


R/get_id_map.R
#' Get a map of different organisation IDs.
#'
#' ROR holds multiple types of unique ID for organisations. This function let's you enter one such ID and return all the other IDs in the database for that organisation. NOTE THIS FUNCTION IS ONLY RECOMMENDED IF YOU HAVE A RELATIVELY SMALL NUMBER OF IDs TO MAP. Otherwise use `read_data_dump()` to access a snapshot of the full dataset.
#' @param id The unique, non-ROR ID for the organisation concerned.
#' @return A list of any available IDs (or a message confirming a match could not be found).
#' @export

get_id_map <- function(id) {

  #convert id into format for entering into query
  fmt_id <- paste0("%22", id, "%22")

  #list of files
  response <- httr::GET(
    url = paste0("https://api.ror.org/organizations?query=", fmt_id),
    timeout = httr::timeout(15)
  )

  #display error message if required
  if(response$status_code >= 400) {
    err_msg = httr::http_status(response)
    stop(err_msg)
  }

  #response text
  response_text <- httr::content(response, "text")

  #json
  response_json <- jsonlite::fromJSON(response_text)

  #checkl if there are actually any results
  if(length(response_json$items) == 0) {

    print("No results found - stopping process")

    return("No IDs found")

    }

  #response df
  response_df <- dplyr::bind_cols(
    response_json[["items"]][["name"]],
    response_json[["items"]][["external_ids"]]
  )

  #column names - use to make sensible names later
  response_colnames <- colnames(response_df)[2:length(colnames(response_df))]

  #new colnames
  new_colnames <- c()

  for (c in response_colnames) {

    new_colnames[2 * match(c, response_colnames) - 1] <- paste0("preferred_", c)
    new_colnames[2 * match(c, response_colnames)] <- paste0("all_", c)

  }

  #unnest response and set new names
  response_final <- response_df |>
    tidyr::unnest(cols = dplyr::everything(), names_repair = "unique") |>
    stats::setNames(c("name", new_colnames)) |>
    dplyr::mutate(dplyr::across(dplyr::everything(), as.character)) |>
    tidyr::pivot_longer(
      cols = -1,
      names_to = "temp",
      values_to = "id"
    ) |>
    tidyr::separate(
      temp,
      into = c("type", "source"),
      sep = "_"
    )



}


R/get_single_org.R
#' Return record for a single specified organisation.
#'
#' Returns the details from ROR API for the organisation specified.
#' @param id The ROR ID of the organisation in question. Can be found using `get_org_list()` or via looking up on the ROR website.
#' @param include_non_active Option to include 'inactive' or 'withdrawn' entities. Must be 'true' or 'false'.
#' @return A list holding the pertinent details for the chosen organisation.
#' @export

get_single_org <- function(id, include_non_active = "false") {


  #run query
  result <- httr::GET(
    url = paste0("https://api.ror.org/organizations/", id, "?all_status=", include_non_active)
  )

  #error result if needed
  if(result$status_code >= 400) {
    err_msg = httr::http_status(result)
    stop(err_msg)
  }

  #text format
  result_text <- httr::content(result, "text")

  #json
  result_json <- jsonlite::fromJSON(result_text)

}


R/read_data_dump.R
#' Read the ROR data dump
#'
#' In addition to their API, ROR also provide a data dump via Zenodo (updated ~quarterly). It is stored in .zip format. This function saves to a temp directory and then turns the .json file within into a dataframe.
#' @param os Defaults to Windows - if using Apple set to "mac" to read correct files format.
#' @return A list of dataframes holding the data dump.
#' @export

read_data_dump <- function(os = "windows") {

  #file url
  zenodo_url <- "https://zenodo.org/api/records/?communities=ror-data&sort=mostrecent"

  #list of files
  response <- httr::GET(
    url = zenodo_url,
    timeout = httr::timeout(15)
  )

  #display error message if required
  if(response$status_code >= 400) {
    err_msg = httr::http_status(response)
    stop(err_msg)
  }

  #result in text format
  response_text <- httr::content(response, "text")

  #resposne df
  file_url <- jsonlite::fromJSON(response_text) |>
    dplyr::filter(created == max(created)) |>
    tidyr::unnest(files, names_repair = "unique") |>
    tidyr::unnest(links...10) |>
    dplyr::pull(download)

  #save to a temp directory
  temp_name <- tempfile()
  curl::curl_download(file_url, temp_name)

  #ui message - reading json
  usethis::ui_info("Reading .json file - may take some time.")

  #list of files
  dd_df <- utils::unzip(temp_name)[ifelse(os == "mac", 2, 1)] |>
    jsonlite::fromJSON()

  #ui message - json read successfully
  usethis::ui_done(".json file read successfully - process completed.")

  #return the data dump
  return(dd_df)



}


R/get_results_by_page.R
#' Helper function - returns org list details one page at a time..
#'
#' Created as a helper function to return org list results one page at a time. Can be used in its own right, however.
#' @param ..page_num Return a specific page number. Enter as an integer - if not set, defaults to 1.
#' @param ..search_terms Query term to narrow results if desired. Separate multiple terms with "|" to search on an OR basis. Will search all indexed fields.
#' @param ..affiliation Specific search term for an affiliation. The API determines closeness of match - only those deemed close enough ('chosen') returned. Will only search name, label and alias fields. You cannot combine affiliation searches with filters.
#' @param ..org_type Only return specified type of organisation. Any of "Education", "Healthcare", "Company", "Archive", "Nonprofit", "Government", "Facility" or "Other".
#' @param ..country_code Filter to specific countries as per the ISO 3166 alpha-2 list. Can combine with `..org_type` filter.
#' @param ..country_name Alternatively, filter by country code instead of name. Can combine with `..org_type` filter.
#' @param ..include_non_active Option to include 'inactive' or 'withdrawn' entities. Must be 'true' or 'false'.
#' @return A dataframe holding the first page of results from the API.
#' @export
#'

get_results_by_page <- function(
  ..page_num = 1,
  ..search_terms,
  ..affiliation,
  ..org_type,
  ..country_code,
  ..country_name,
  ..include_non_active = "false"
)  {

  #sleep for 1 second to avoid throttling when use in `get_org_list()`
  Sys.sleep(1)

  #hack to deal with missing errors
  ot_hack <- if(missing(..org_type)) {"blah"} else {..org_type}
  cc_hack <- if(missing(..country_code)) {"blah"} else {..country_code}
  cn_hack <- if(missing(..country_name)) {"blah"} else {..country_name}

  #error handling----------------------------------------------

  #type
  if(!missing(..org_type) & !ot_hack %in% c("Education", "Healthcare", "Company", "Archive", "Nonprofit", "Government", "Facility", "Other"))
    stop("`.org_type` must be (if used) one of: 'Education', 'Healthcare', 'Company', 'Archive', 'Nonprofit', 'Government', 'Facility', or 'Other'.")

  #only one org type, country code, or country
  if(
    (!missing(..org_type) & length(ot_hack) > 1) |
    (!missing(..country_code) & length(cc_hack) > 1) |
    (!missing(..country_name) & length(cn_hack) > 1)
  ) stop("You can only enter a single filter value for each parameter.")

  #only filter one of country_code or country_name
  if(!missing(..country_code) & !missing(..country_name))
    stop("You can only filter one of `.country_code` or `.country_name`.")

  #only search affiliation if chosen
  if(!missing(..affiliation) & (!missing(..search_terms) | !missing(..org_type) | !missing(..country_code) | !missing(..country_name)))
    stop("If using `..affiliation` you can not use `.search_terms` or any filters.")

  #non-active must be true or false
  if(
    !..include_non_active %in% c("true", "false")
  ) stop("`.include_non_active` must be one of 'true' or 'false'.")

  #turn spaces into "%20" for upload-------------------------------------------------

  #affiliation
  if(!missing(..affiliation)) {..affiliation <- gsub(" ", "%20", ..affiliation)}

  #search terms
  if(!missing(..search_terms)) {..search_terms <- gsub(" ", "%20", ..search_terms)}

  #country name
  if(!missing(..country_name)) {..country_name <- gsub(" ", "%20", ..country_name)}

  #query settings for prelimianry run (to identify number of pages)-----------------------------------------------

  #affiliation entered
  if(!missing(..affiliation)) {

      .query_settings <- list(
        page = ..page_num,
        affiliation = ..affiliation,
        active_status = ..include_non_active
        )

      #no parameters (except maybe page number)
  } else if (
    missing(..search_terms) &
    missing(..org_type) &
    missing(..affiliation) &
    missing(..country_code) &
    missing(..country_name)
  ) {

    .query_settings <- list(
      page = ..page_num,
      all_status = ..include_non_active
    )

    #search term (plus possibly page_number) only
    } else if(!missing(..search_terms) & missing(..org_type) & missing(..country_code) & missing(..country_name)) {

  .query_settings <- list(
    page = ..page_num,
    all_status = ..include_non_active,
    query = ..search_terms
  )

  #search term and org_type (plus possibly page_number)
} else if(!missing(..search_terms) & !missing(..org_type) & missing(..country_code) & missing(..country_name)) {

  .query_settings <- list(
    page = ..page_num,
    query = ..search_terms,
    all_status = ..include_non_active,
    filter = paste0("types:", ..org_type)
  )

  #search term, org_type and country code
} else if(!missing(..search_terms) & !missing(..org_type) & !missing(..country_code)) {

  .query_settings <- list(
    page = ..page_num,
    query = ..search_terms,
    all_status = ..include_non_active,
    filter = paste0("types:", ..org_type, ",country.country_code:", ..country_code)
  )

  #search term, org_type and country name
} else if(!missing(..search_terms) & !missing(..org_type) & !missing(..country_name)) {

  .query_settings <- list(
    page = ..page_num,
    query = ..search_terms,
    all_status = ..include_non_active,
    filter = paste0("types:", ..org_type, ",country.country_name:", ..country_name)
  )

  #org type only
} else if(missing(..search_terms) & !missing(..org_type) & missing(..country_code) & missing(..country_name)) {

  .query_settings <- list(
    page = ..page_num,
    all_status = ..include_non_active,
    filter = paste0("types:", ..org_type)
  )

  #org_type and country code
}  else if(missing(..search_terms) & !missing(..org_type) & !missing(..country_code)) {

  .query_settings <- list(
    page = ..page_num,
    all_status = ..include_non_active,
    filter = paste0("types:", ..org_type, ",country.country_code:", ..country_code)
  )

  #org_type and country name
} else if(missing(..search_terms) & !missing(..org_type) & !missing(..country_name)) {

  .query_settings <- list(
    page = ..page_num,
    all_status = ..include_non_active,
    filter = paste0("types:", ..org_type, ",country.country_name:", ..country_name)
  )

  #country code only
} else if(missing(..search_terms) & missing(..org_type) & !missing(..country_code)) {

  .query_settings <- list(
    page = ..page_num,
    all_status = ..include_non_active,
    filter = paste0("country.country_code:", ..country_code)
  )

  #country name only
} else if(missing(..search_terms) & missing(..org_type) & !missing(..country_name)) {

  .query_settings <- list(
    page = ..page_num,
    all_status = ..include_non_active,
    filter = paste0("country.country_name:", ..country_name)
  )

} else if(!missing(..search_terms) & missing(..org_type) & !missing(..country_code)) {

  .query_settings <- list(
    page = 1,
    query = .search_terms,
    all_status = ..include_non_active,
    filter = paste0("country.country_code:", .country_code)
  )

  #search terms and country name
} else if(!missing(..search_terms) & missing(..org_type) & !missing(..country_name)) {

  .query_settings <- list(
    page = 1,
    query = ..search_terms,
    filter = paste0("country.country_name:", ..country_name)
  )

}

#run query--------------------------------------

#try running query
result <- httr::GET(
  url = paste0("https://api.ror.org/organizations"),
  query = .query_settings,
  timeout = httr::timeout(15)
)

#display error message if required
if(result$status_code >= 400) {
  err_msg = httr::http_status(result)
  stop(err_msg)
}

#results of query-=------------------

#text
result_text <- httr::content(result, "text")

#json
result_json <- jsonlite::fromJSON(result_text)

#final results, depending on whether affiliation used or not
if(missing(..affiliation)){

  result_df <- result_json[["items"]]

} else {

  results_df <- result_json[["items"]] |>
    dplyr::filter(chosen == T) |>
    tidyr::unnest(organization)

}

}


R/get_org_list.R
#' Query organisation records to return a list; you can add query terms to refine results.
#'
#' By defaults returns all results from the API. Also let's you specify a specific page number, filter your results by specific categories, or add search terms to your query to refine the results. It is HIGHLY RECOMMENDED you filter your results in some way to minimise running times.
#' @param .search_terms Query term to narrow results if desired. Separate multiple terms with "|" to search on an OR basis. Will search all indexed fields.
#' @param .affiliation Specific search term for an affiliation. The API determines closeness of match - only those deemed close enough ('chosen') returned. Will only search name, label and alias fields. You cannot combine affiliation searches with filters.
#' @param .org_type Only return specified type of organisation. Any of "Education", "Healthcare", "Company", "Archive", "Nonprofit", "Government", "Facility" or "Other".
#' @param .country_code Filter to specific countries as per the ISO 3166 alpha-2 list. Can combine with `.org_type` filter.
#' @param .country_name Alternatively, filter by country code instead of name. Can combine with `.org_type` filter.
#' @param .page_numbers Return a specific page number (or set of page numbers). Enter as a numeric vector. All results will be returned if you do not specify. If you specify page numbers that are not found, all results will be returned.
#' @param .include_non_active Option to include 'inactive' or 'withdrawn' entities. Must be 'true' or 'false'.
#' @return A dataframe holding results as per specifications (list of all orgs returned if no parameters are set).
#' @export

get_org_list <- function(
  .search_terms,
  .affiliation,
  .org_type,
  .country_code,
  .country_name,
  .page_numbers,
  .include_non_active = "false"
) {

  #hack to deal with missing errors
  ot_hack <- if(missing(.org_type)) {"blah"} else {.org_type}
  cc_hack <- if(missing(.country_code)) {"blah"} else {.country_code}
  cn_hack <- if(missing(.country_name)) {"blah"} else {.country_name}

  #error handling----------------------------------------------

  #type
  if(!missing(.org_type) & !ot_hack %in% c("Education", "Healthcare", "Company", "Archive", "Nonprofit", "Government", "Facility", "Other"))
    stop("`.org_type` must be (if used) one of: 'Education', 'Healthcare', 'Company', 'Archive', 'Nonprofit', 'Government', 'Facility', or 'Other'.")

  #only one org type, country code, or country
  if(
    (!missing(.org_type) & length(ot_hack) > 1) |
    (!missing(.country_code) & length(cc_hack) > 1) |
    (!missing(.country_name) & length(cn_hack) > 1)
  ) stop("You can only enter a single filter value for each parameter.")

  #only filter one of country_code or country_name
  if(!missing(.country_code) & !missing(.country_name))
    stop("You can only filter one of `.country_code` or `.country_name`.")

  #only search affiliation if chosen
  if(!missing(.affiliation) & (!missing(.search_terms) | !missing(.org_type) | !missing(.country_code) | !missing(.country_name)))
    stop("If using `.affiliation` you can not use `.search_terms` or any filters.")

  #non-active must be true or false
  if(
    !.include_non_active %in% c("true", "false")
  ) stop("`.include_non_active` must be one of 'true' or 'false'.")


  #turn spaces into "%20" for upload-------------------------------------------------

  #affiliation
  if(!missing(.affiliation)) {.affiliation <- gsub(" ", "%20", .affiliation)}

  #search terms
  if(!missing(.search_terms)) {.search_terms <- gsub(" ", "%20", .search_terms)}

  #country name
  if(!missing(.country_name)) {.country_name <- gsub(" ", "%20", .country_name)}

  #query settings for prelimianry run (to identify number of pages)-----------------------------------------------

  #no paramaters set (or only page numbers)
  if(
    missing(.search_terms) &
    missing(.affiliation) &
    missing(.org_type) &
    (missing(.country_code) | missing(.country_name))
  ) {

    query_settings <- list(
      page = 1,
      all_status = .include_non_active
    )

    #affiliation entered
  } else if(!missing(.affiliation)) {

    query_settings <- list(
      page = 1,
      affiliation = .affiliation,
      all_status = .include_non_active
    )

    #search term (plus possibly page_number) only
  } else if(!missing(.search_terms) & missing(.org_type) & missing(.country_code) & missing(.country_name)) {

    query_settings <- list(
      page = 1,
      query = .search_terms,
      all_status = .include_non_active
    )

    #search term and org_type (plus possibly page_number)
  } else if(!missing(.search_terms) & !missing(.org_type) & missing(.country_code) & missing(.country_name)) {

    query_settings <- list(
      page = 1,
      query = .search_terms,
      all_status = .include_non_active,
      filter = paste0("types:", .org_type)
    )

    #search term, org_type and country code
  } else if(!missing(.search_terms) & !missing(.org_type) & !missing(.country_code)) {

    query_settings <- list(
      page = 1,
      query = .search_terms,
      all_status = .include_non_active,
      filter = paste0("types:", .org_type, ",country.country_code:", .country_code)
    )

    #search term, org_type and country name
  } else if(!missing(.search_terms) & !missing(.org_type) & !missing(.country_name)) {

    query_settings <- list(
      page = 1,
      query = .search_terms,
      all_status = .include_non_active,
      filter = paste0("types:", .org_type, ",country.country_name:", .country_name)
    )

    #org type only
  } else if(missing(.search_terms) & !missing(.org_type) & missing(.country_code) & missing(.country_name)) {

    query_settings <- list(
      page = 1,
      all_status = .include_non_active,
      filter = paste0("types:", .org_type)
    )

    #org_type and country code
  }  else if(missing(.search_terms) & !missing(.org_type) & !missing(.country_code)) {

    query_settings <- list(
      page = 1,
      all_status = .include_non_active,
      filter = paste0("types:", .org_type, ",country.country_code:", .country_code)
    )

    #org_type and country name
  } else if(missing(.search_terms) & !missing(.org_type) & !missing(.country_name)) {

    query_settings <- list(
      page = 1,
      all_status = .include_non_active,
      filter = paste0("types:", .org_type, ",country.country_name:", .country_name)
    )

    #country code only
  } else if(missing(.search_terms) & missing(.org_type) & !missing(.country_code)) {

    query_settings <- list(
      page = 1,
      all_status = .include_non_active,
      filter = paste0("country.country_code:", .country_code)
    )

    #country name only
  } else if(missing(.search_terms) & missing(.org_type) & !missing(.country_name)) {

    query_settings <- list(
      page = 1,
      all_status = .include_non_active,
      filter = paste0("country.country_name:", .country_name)
    )

    #search term and country code
  } else if(!missing(.search_terms) & missing(.org_type) & !missing(.country_code)) {

    query_settings <- list(
      page = 1,
      query = .search_terms,
      all_status = .include_non_active,
      filter = paste0("country.country_code:", .country_code)
    )

    #search terms and country name
  } else if(!missing(.search_terms) & missing(.org_type) & !missing(.country_name)) {

    query_settings <- list(
      page = 1,
      query = .search_terms,
      all_status = .include_non_active,
      filter = paste0("country.country_name:", .country_name)
    )

  }

  #run preliminary query--------------------------------------

  #try running query
  prelim_result <- httr::GET(
    url = paste0("https://api.ror.org/organizations"),
    query = query_settings,
    timeout = httr::timeout(15)
  )

  #display error message if required
  if(prelim_result$status_code >= 400) {
    err_msg = httr::http_status(prelim_result)
    stop(err_msg)
  }

  #results of preliminary query - use to determine page numbers if required-=------------------

  #text
  prelim_result_text <- httr::content(prelim_result, "text")

  #json
  prelim_result_json <- jsonlite::fromJSON(prelim_result_text)

  #total pages
  total_pages <- ceiling((prelim_result_json$number_of_results / 20))

  #determine page numbers
  if(missing(.page_numbers)) {

    #pages bector
    pages <- c(1:total_pages)

  } else if (any(.page_numbers > total_pages)) {

    pages <- c(1:total_pages)

  } else {

    pages <- .page_numbers

    }

  #get all required results-------------------------------------------

  #affiliation parameter
  if(!missing(.affiliation)) {


    result_df <- lapply(
      X = pages,
      FUN = get_results_by_page,
      ..affiliation = .affiliation,
      ..include_non_active = .include_non_active
    ) |>
      dplyr::bind_rows()

    #no parameters
  } else if(
    missing(.affiliation) &
    missing(.org_type) &
    missing(.country_code) &
    missing(.country_name) &
    missing(.search_terms)
    ) {

    result_df <- lapply(
      X = pages,
      FUN = get_results_by_page,
      ..include_non_active = .include_non_active
    ) |>
      dplyr::bind_rows()


    #search term only
  } else if(!missing(.search_terms) & missing(.org_type) & missing(.country_code) & missing(.country_name)) {

    result_df <- lapply(
      X = pages,
      FUN = get_results_by_page,
      ..search_terms = .search_terms,
      ..include_non_active = .include_non_active
    ) |>
      dplyr::bind_rows()


    #search term and org type
  } else if(!missing(.search_terms) & !missing(.org_type) & missing(.country_code) & missing(.country_name)) {

    result_df <- lapply(
      X = pages,
      FUN = get_results_by_page,
      ..search_terms = .search_terms,
      ..include_non_active = .include_non_active,
      ..org_type = .org_type
    ) |>
      dplyr::bind_rows()

    #search term, org_type and country code
  } else if(!missing(.search_terms) & !missing(.org_type) & !missing(.country_code)) {

    result_df <- lapply(
      X = pages,
      FUN = get_results_by_page,
      ..search_terms = .search_terms,
      ..include_non_active = .include_non_active,
      ..org_type = .org_type,
      ..country_code = .country_code
    ) |>
      dplyr::bind_rows()


    #search term, org type and country name
  } else if(!missing(.search_terms) & !missing(.org_type) & !missing(.country_name)) {

    result_df <- lapply(
      X = pages,
      FUN = get_results_by_page,
      ..search_terms = .search_terms,
      ..include_non_active = .include_non_active,
      ..org_type = .org_type,
      ..country_name = .country_name
    ) |>
      dplyr::bind_rows()


    #org type only
  } else if(missing(.search_terms) & !missing(.org_type) & missing(.country_code) & missing(.country_name)) {

    result_df <- lapply(
      X = pages,
      FUN = get_results_by_page,
      ..include_non_active = .include_non_active,
      ..org_type = .org_type
    ) |>
      dplyr::bind_rows()

    #org type and country code
  }  else if(missing(.search_terms) & !missing(.org_type) & !missing(.country_code)) {

    result_df <- lapply(
      X = pages,
      FUN = get_results_by_page,
      ..org_type = .org_type,
      ..include_non_active = .include_non_active,
      ..country_code = .country_code
    ) |>
      dplyr::bind_rows()

    #org type and coutry name
  } else if(missing(.search_terms) & !missing(.org_type) & !missing(.country_name)) {

    result_df <- lapply(
      X = pages,
      FUN = get_results_by_page,
      ..org_type = .org_type,
      ..include_non_active = .include_non_active,
      ..country_name = .country_name
    ) |>
      dplyr::bind_rows()

    #country code only
  } else if(missing(.search_terms) & missing(.org_type) & !missing(.country_code)) {

    result_df <- lapply(
      X = pages,
      FUN = get_results_by_page,
      ..country_code = .country_code,
      ..include_non_active = .include_non_active
    ) |>
      dplyr::bind_rows()


    #country nasme only
  } else if(missing(.search_terms) & missing(.org_type) & !missing(.country_name)) {

    result_df <- lapply(
      X = pages,
      FUN = get_results_by_page,
      ..country_name = .country_name,
      ..include_non_active = .include_non_active
    ) |>
      dplyr::bind_rows()

    #search term and country code
  } else if(!missing(.search_terms) & missing(.org_type) & !missing(.country_code)) {

    result_df <- lapply(
      X = pages,
      FUN = get_results_by_page,
      ..search_terms = .search_terms,
      ..country_code = .country_code,
      ..include_non_active = .include_non_active
    ) |>
      dplyr::bind_rows()


  } else if(!missing(.search_terms) & missing(.org_type) & !missing(.country_name)) {

    result_df <- lapply(
      X = pages,
      FUN = get_results_by_page,
      ..search_terms = .search_terms,
      ..country_name = .country_name,
      ..include_non_active = .include_non_active
    ) |>
      dplyr::bind_rows()


  }


return(result_df)



}
